coderd/agentapi/stats_test.go

```diff
@@ -0,0 +1,375 @@
+				},
+				SetFn: func(context.Context, database.Store, database.Template, schedule.TemplateScheduleOptions) (database.Template, error) {
+					panic("not implemented")
+				},
+			}
+			batcher = &statsBatcher{}
+
+			req = &agentproto.UpdateStatsRequest{
+				Stats: &agentproto.Stats{
+					ConnectionsByProto: map[string]int64{
+						"tcp": 1,
+					},
+					ConnectionCount:           0,
+					ConnectionMedianLatencyMs: 23,
+				},
+			}
+		)
+		api := agentapi.StatsAPI{
+			AgentFn: func(context.Context) (database.WorkspaceAgent, error) {
+				return agent, nil
+			},
+			Database:                  dbM,
+			StatsBatcher:              batcher,
+			TemplateScheduleStore:     templateScheduleStorePtr(templateScheduleStore),
+			AgentStatsRefreshInterval: 10 * time.Second,
+			// Ignored when nil.
+			UpdateAgentMetricsFn: nil,
+			TimeNowFn: func() time.Time {
+				return now
+			},
+		}
+
+		// Workspace gets fetched.
+		dbM.EXPECT().GetWorkspaceByAgentID(gomock.Any(), agent.ID).Return(database.GetWorkspaceByAgentIDRow{
+			Workspace:    workspace,
+			TemplateName: template.Name,
+		}, nil)
+
+		// Workspace last used at gets bumped.
+		dbM.EXPECT().UpdateWorkspaceLastUsedAt(gomock.Any(), database.UpdateWorkspaceLastUsedAtParams{
+			ID:         workspace.ID,
+			LastUsedAt: now,
+		}).Return(nil)
+
+		_, err := api.UpdateStats(context.Background(), req)
+		require.NoError(t, err)
+	})
+
+	t.Run("NoConnectionsByProto", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			dbM = dbmock.NewMockStore(gomock.NewController(t))
+			req = &agentproto.UpdateStatsRequest{
+				Stats: &agentproto.Stats{
+					ConnectionsByProto: map[string]int64{}, // len() == 0
+				},
+			}
+		)
+		api := agentapi.StatsAPI{
+			AgentFn: func(context.Context) (database.WorkspaceAgent, error) {
+				return agent, nil
+			},
+			Database:                  dbM,
+			StatsBatcher:              nil, // should not be called
+			TemplateScheduleStore:     nil, // should not be called
+			AgentStatsRefreshInterval: 10 * time.Second,
+			UpdateAgentMetricsFn:      nil, // should not be called
+			TimeNowFn: func() time.Time {
+				panic("should not be called")
+			},
+		}
+
+		resp, err := api.UpdateStats(context.Background(), req)
+		require.NoError(t, err)
+		require.Equal(t, &agentproto.UpdateStatsResponse{
+			ReportInterval: durationpb.New(10 * time.Second),
+		}, resp)
+	})
+
+	t.Run("AutostartAwareBump", func(t *testing.T) {
+		t.Parallel()
+
+		// Use a workspace with an autostart schedule.
+		workspace := workspace
+		workspace.AutostartSchedule = sql.NullString{
+			String: "CRON_TZ=Australia/Sydney 0 8 * * *",
+			Valid:  true,
+		}
+
+		// Use a custom time for now which would trigger the autostart aware
+		// bump.
+		now, err := time.Parse("2006-01-02 15:04:05 -0700 MST", "2023-12-19 07:30:00 +1100 AEDT")
+		require.NoError(t, err)
+		now = dbtime.Time(now)
+		nextAutostart := now.Add(30 * time.Minute).UTC() // always sent to DB as UTC
```

Comment:
to me, this smells like we need to refactor.  The AgentAPI should not be in the business of computing autostart times.  There should be a scheduler component that gets an activity indication from the Agent API and it computes and decides how to update the autostart / autostop schedule.

Another one for some other PR...